File: Godeps/_workspace/src/github.com/cockroachdb/c-rocksdb/internal/java/src/main/java/org/rocksdb/Statistics.java
// Copyright (c) 2014, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.

package org.rocksdb;

/**
 * Statistics to analyze the performance of a db. Pointer for statistics object
 * is managed by Options class.
 */
public class Statistics {

  private final long statsHandle_;

  public Statistics(final long statsHandle) {
    statsHandle_ = statsHandle;
  }

  public long getTickerCount(TickerType tickerType) {
    assert(isInitialized());
    return getTickerCount0(tickerType.getValue(), statsHandle_);
  }

  public HistogramData geHistogramData(final HistogramType histogramType) {
    assert(isInitialized());
    return geHistogramData0(
        histogramType.getValue(), statsHandle_);
  }

  private boolean isInitialized() {
    return (statsHandle_ != 0);
  }

  private native long getTickerCount0(int tickerType, long handle);
  private native HistogramData geHistogramData0(int histogramType, long handle);
}

File: Godeps/_workspace/src/github.com/cockroachdb/c-rocksdb/internal/java/rocksjni/statistics.cc
// Copyright (c) 2014, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.
//
// This file implements the "bridge" between Java and C++ and enables
// calling c++ rocksdb::Statistics methods from Java side.

#include <stdio.h>
#include <stdlib.h>
#include <jni.h>

#include "include/org_rocksdb_Statistics.h"
#include "rocksjni/portal.h"
#include "rocksdb/statistics.h"

/*
 * Class:     org_rocksdb_Statistics
 * Method:    getTickerCount0
 * Signature: (IJ)J
 */
jlong Java_org_rocksdb_Statistics_getTickerCount0(
    JNIEnv* env, jobject jobj, int tickerType, jlong handle) {
  auto st = reinterpret_cast<rocksdb::Statistics*>(handle);
  assert(st != nullptr);

  return st->getTickerCount(static_cast<rocksdb::Tickers>(tickerType));
}

/*
 * Class:     org_rocksdb_Statistics
 * Method:    geHistogramData0
 * Signature: (IJ)Lorg/rocksdb/HistogramData;
 */
jobject Java_org_rocksdb_Statistics_geHistogramData0(
  JNIEnv* env, jobject jobj, int histogramType, jlong handle) {
  auto st = reinterpret_cast<rocksdb::Statistics*>(handle);
  assert(st != nullptr);

  rocksdb::HistogramData data;
  st->histogramData(static_cast<rocksdb::Histograms>(histogramType),
    &data);

  // Don't reuse class pointer
  jclass jclazz = env->FindClass("org/rocksdb/HistogramData");
  jmethodID mid = rocksdb::HistogramDataJni::getConstructorMethodId(
      env, jclazz);
  return env->NewObject(jclazz, mid, data.median, data.percentile95,
      data.percentile99, data.average, data.standard_deviation);
}
